ln-gateway/src/ln.rs

```diff
@@ -66,3 +70,66 @@ impl LnRpc for Mutex<cln_rpc::ClnRpc> {
         }
     }
 }
+
+/// Control the behavoir of the LnRpc by exposing the same interface but attach configurations and
+/// methods to it.
+pub struct LnRpcAdapter {
+    // add more configuration
+    ln_client: Box<dyn LnRpc>,
+    fail_times: u8,
+    fail_invoice: Arc<Mutex<HashMap<String, AtomicU8>>>, //does this need to be in arc
+}
+
+impl LnRpcAdapter {
+    pub fn new(ln_client: Box<dyn LnRpc>, fail_times: u8) -> Self {
+        let fail_invoice = Arc::new(Mutex::new(HashMap::new()));
+
+        LnRpcAdapter {
+            ln_client,
+            fail_times,
+            fail_invoice,
+        }
+    }
+
+    /// Tell the LnRpc to fail the payment x-times until succeeding (or trying to succeed in the
+    /// case of real LnRpc
+    pub async fn fail_invoice(&self, invoice: String, times: u8) {
+        self.fail_invoice
+            .lock()
+            .await
+            .insert(invoice, AtomicU8::new(times + 1));
+    }
+}
+
+#[async_trait]
+impl LnRpc for LnRpcAdapter {
+    async fn pay(
+        &self,
+        invoice_str: &str,
+        max_delay: u64,
+        max_fee_percent: f64,
+    ) -> Result<[u8; 32], LightningError> {
+        self.fail_invoice
+            .lock()
+            .await
+            .entry(invoice_str.to_string())
+            .and_modify(|counter| {
+                let _ = counter.fetch_sub(1, Ordering::Relaxed);
+            })
+            .or_insert_with(|| AtomicU8::new(self.fail_times));
+        if self
+            .fail_invoice
+            .lock()
+            .await
+            .get(invoice_str)
+            .expect("we always either already have an invoice or insert an invoice")
+            .load(Ordering::Relaxed)
```

Comment:
I don't care about perf, but throwing the result of `fetch_sub` just to repeat the whole locking and `load` is a lot of extra code.

Also -I don't think it handles wrapping around correctly now. It always decrements, even when `0` already.

Reply:
true I'll change that

Reply:
(not using atomics anymore)